### ISSUE TITLE
Error: nvml: Driver/library version mismatch

### ISSUE THREAD

**Issue by sandipmgiri** [NVIDIA/nvidia-docker#584]
Initially `nividia-docker` and `nvidia-smi ` both the commands were working fine. Now I ran "yum update -y" on my RHEL vm , and looks like some drivers are updated , currently I'm getting following error for `nvidia-smi`  : 

```
$ nvidia-smi
Failed to initialize NVML: Function Not Found
```

Also tried to run the nvidia-docker : 
```
$ sudo nvidia-docker run -it nvidia/cuda-ppc64le:8.0-cudnn6-devel-ubuntu16.04 nvidia-smi
nvidia-docker | 2017/12/21 05:04:46 Error: nvml: Driver/library version mismatch
```

Any help appreciated .Thanks!

**Comment by 3XX0** [comment]
You probably need to reboot or reload the `nvidia*` kernel modules

**Comment by sandipmgiri** [comment]
Rebooted , but getting same issue.

**Comment by pineking** [comment]
@sandipmgiri try to re-install the drivers and nvidia-docker

**Comment by sandipmgiri** [comment]
The system is back up and running.  There were symbolic links to old revs of the driver that needed to be updated in /usr/lib64/nvidia  to the libnvidia-ml* files associated with the latest driver loaded with CUDA 9.1.

**Comment by cipri-tom** [comment]
@sandipmgiri did you have to update them by hand, or they were resolved by reboot ? Also, are these revs of the driver on the machine or inside docker ?

Thank you !

**Comment by sandipmgiri** [comment]
I have updated manually on the machine, see below command :
sudo ln -s /usr/local/nvidia-driver/nvidia_driver/387.36/lib64/libnvidia-ml.so.1 /usr/lib64/nvidia

**Comment by barzan-hayati** [comment]
>  sudo ln -s /usr/local/nvidia-driver/nvidia_driver/387.36/lib64/libnvidia-ml.so.1 /usr/lib64/nvidia

I check your command :
`ln: failed to create symbolic link '/usr/lib64/nvidia': No such file or directory`

**Comment by ravikiranrao** [comment]
Same issue. Anyone has a solution.

@barzan-hayati I too got the same error.

**Comment by pbruneau** [comment]
The comment from @sandipmgiri gave me the lead to the definitive answer IMHO.

**tl;dr:**

- Update your host with the latest stable Nvidia drivers / CUDA combination (e.g. 440.33 and CUDA 10.2 for Ubuntu 16.04, 450.51 and CUDA 11.0 update 1 for Ubuntu 18.04), as set by the `.deb` CUDA installers obtained from Nvidia download site
- Then containers for any `nvidia/cuda:xxxx-ubuntuyyyy` (with yyyy the system version on your host) should run fine.

**The reason:**

I think the origin comes from inconstencies when using `nvidia/cuda:xxxx-ubuntuyyyy` images from Dockerhub. For example, when running a container derived from `nvidia/cuda:10.0-cudnn7-devel-ubuntu18.04`, whatever you do, some libraries will link to the 450.51 Nvidia driver version (the reference driver version for CUDA 11.0 update 1 / CUDNN 8), instead of 410.129 (the driver for CUDA 10.0 / CUDNN 7), irrespective of the configuration of your host. For example, from the container:

```
ls -ltar /usr/lib/x86_64-linux-gnu/libnvidia-ptx*
-rw-r--r-- 1 root root 12129448 Aug 20  2019 /usr/lib/x86_64-linux-gnu/libnvidia-ptxjitcompiler.so.410.129
-rw-r--r-- 1 root root  9947144 Jul 19 20:12 /usr/lib/x86_64-linux-gnu/libnvidia-ptxjitcompiler.so.450.51.06
lrwxrwxrwx 1 root root       37 Jul 20 00:24 /usr/lib/x86_64-linux-gnu/libnvidia-ptxjitcompiler.so.1 -> libnvidia-ptxjitcompiler.so.450.51.06
lrwxrwxrwx 1 root root       29 Jul 20 00:24 /usr/lib/x86_64-linux-gnu/libnvidia-ptxjitcompiler.so -> libnvidia-ptxjitcompiler.so.1
```

We see libraries are present for 2 driver versions (410.129, which works for CUDA 10.0, that I want to use, and that I was setting on my host in order to try to be consistent, and 450.51, which works for CUDA 11.0), but the symbolic link is made to the Ubuntu latest stable Nvidia/CUDA combination, instead of the one you want to use in your container. Fixing with `ln -s` does not work in a persistent way, as links are restored as before at container restart.

So you have to make sure your host runs the latest stable CUDA and associated drivers for your system (i.e. for Ubuntu 16.04, CUDA 10.2 and 440.33, for Ubuntu 18.04, CUDA 11.0 and 450.51). Then you can run a container with any CUDA/CUDNN version you want (in my case above, 10.0), and it runs just fine.
